File: backend/src/kuuna_backend/api/schemas/gateway.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field


GatewayEventType = Literal["message_created", "message_edited", "message_deleted"]
GatewayProvider = Literal["whatsapp-neonize"]
OutboundDeliveryStatus = Literal["sent", "failed", "retrying"]


class InboundMedia(BaseModel):
    provider_media_id: str
    mime_type: str
    file_name: str | None = None
    byte_size: int | None = None
    download_url: str | None = None


class InboundMessage(BaseModel):
    text: str | None = None
    reply_to_provider_message_id: str | None = None
    mentions: list[str] = Field(default_factory=list)
    media: list[InboundMedia] = Field(default_factory=list)


class GatewayInboundEvent(BaseModel):
    trace_id: UUID
    provider: GatewayProvider
    provider_group_id: str
    provider_message_id: str
    sender_provider_user_id: str | None = None
    event_type: GatewayEventType
    occurred_at: datetime
    message: InboundMessage
    raw_event: dict[str, Any] | None = None


class GatewayInboundAccepted(BaseModel):
    accepted: bool
    trace_id: UUID
    deduped: bool = False


class OutboundMetadata(BaseModel):
    agent_instance_id: UUID
    model_path: list[str]


class GatewayOutboundIntent(BaseModel):
    trace_id: UUID
    outbound_intent_id: UUID
    provider_group_id: str
    reply_to_provider_message_id: str | None = None
    text: str
    metadata: OutboundMetadata


class GatewayOutboundStatusEvent(BaseModel):
    trace_id: UUID
    outbound_intent_id: UUID
    status: OutboundDeliveryStatus
    provider_message_id: str | None = None
    error_code: str | None = None
    error_message: str | None = None
    occurred_at: datetime


class GatewayOutboundStatusAccepted(BaseModel):
    accepted: bool
    trace_id: UUID

File: packages/contracts/src/gateway.ts
export type GatewayEventType =
  | "message_created"
  | "message_edited"
  | "message_deleted";

export interface GatewayInboundMedia {
  provider_media_id: string;
  mime_type: string;
  file_name?: string | null;
  byte_size?: number | null;
  download_url?: string | null;
  inline_data_base64?: string | null;
}

export interface GatewayInboundMessage {
  text?: string | null;
  reply_to_provider_message_id?: string | null;
  mentions: string[];
  media: GatewayInboundMedia[];
}

export interface GatewayInboundEvent {
  trace_id: string;
  provider: "whatsapp-neonize";
  provider_group_id: string;
  provider_message_id: string;
  sender_provider_user_id?: string | null;
  event_type: GatewayEventType;
  occurred_at: string;
  message: GatewayInboundMessage;
  raw_event?: Record<string, unknown> | null;
}

export interface GatewayOutboundIntent {
  trace_id: string;
  outbound_intent_id: string;
  provider_group_id: string;
  reply_to_provider_message_id?: string | null;
  text: string;
  metadata: {
    agent_instance_id: string;
    model_path: string[];
  };
}

export interface GatewayOutboundStatusEvent {
  trace_id: string;
  outbound_intent_id: string;
  status: "sent" | "failed" | "retrying";
  provider_message_id?: string | null;
  error_code?: string | null;
  error_message?: string | null;
  occurred_at: string;
}
